sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -147,43 +172,59 @@ private void drainQueue() {
         long numberRequested = REQUESTED.get(this);
         boolean isEmpty = bufferMessages.isEmpty();
 
-        SynchronousReceiveWork work;
+        SynchronousReceiveWork currentDownstream = null;
         while (numberRequested != 0L && !isEmpty) {
             if (isTerminated()) {
                 break;
             }
 
-            long numberEmitted = 0L;
-            while (numberRequested != numberEmitted) {
+            // Track the number of messages read from the buffer those are either emitted to downstream or released.
+            long numberConsumed = 0L;
+            // Iterate and attempt to read the requested number of events.
+            while (numberRequested != numberConsumed) {
                 if (isEmpty || isTerminated()) {
                     break;
                 }
 
                 final ServiceBusReceivedMessage message = bufferMessages.poll();
+
+                // While there are messages in the buffer, obtain the current (unterminated) downstream and
+                // attempt to emit the message to it.
                 boolean isEmitted = false;
                 while (!isEmitted) {
-                    work = getOrUpdateCurrentWork();
-                    if (work == null) {
+                    currentDownstream = getOrUpdateCurrentWork();
+                    if (currentDownstream == null) {
                         break;
                     }
 
-                    isEmitted = work.emitNext(message);
+                    isEmitted = currentDownstream.emitNext(message);
                 }
 
-                // We could not emit the last message that we polled because there were no work items.
-                // Push this back to the head of the work queue.
                 if (!isEmitted) {
-                    bufferMessages.addFirst(message);
-                    break;
+                    // The only reason we can't emit was the downstream(s) were terminated hence nobody
+                    // to receive the message.
+                    if (isPrefetchDisabled) {
+                        // release is enabled only for no-prefetch scenario.
+                        asyncClient.release(message).subscribe(__ -> { },
+                            error -> logger.warning("lockToken[{}] Couldn't release the message.",
```

Comment:
Do we recommend using `logger.atWarning().addKeyValue("lockToken", <>)`?

Reply:
I think lock-token is something eligible to define as an attribute to lookup, but we haven't migrated the SB package to structured logging yet, so I think we should decide the standard attributes (to use as key-value pairs) when making that effort in the SB package as a whole.